File: libs/ml-kernel/src/models/prediction.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FeatureContribution {
    pub name: String,
    pub value: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PredictionOutput {
    pub record_id: String,
    pub variant: String,
    pub model_version_id: Uuid,
    pub predicted_label: String,
    pub score: f64,
    pub confidence: f64,
    pub contributions: Vec<FeatureContribution>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RealtimePredictionRequest {
    #[serde(default)]
    pub inputs: Vec<Value>,
    #[serde(default)]
    pub explain: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RealtimePredictionResponse {
    pub deployment_id: Uuid,
    pub outputs: Vec<PredictionOutput>,
    pub predicted_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateBatchPredictionRequest {
    pub deployment_id: Uuid,
    #[serde(default)]
    pub records: Vec<Value>,
    pub output_destination: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BatchPredictionJob {
    pub id: Uuid,
    pub deployment_id: Uuid,
    pub status: String,
    pub record_count: i64,
    pub output_destination: Option<String>,
    pub outputs: Vec<PredictionOutput>,
    pub created_at: DateTime<Utc>,
    pub completed_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListBatchPredictionsResponse {
    pub data: Vec<BatchPredictionJob>,
}

File: libs/ml-kernel-go/models/prediction.go
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// FeatureContribution is one feature's contribution to a prediction.
type FeatureContribution struct {
	Name  string  `json:"name"`
	Value float64 `json:"value"`
}

// PredictionOutput is one row of a prediction batch.
type PredictionOutput struct {
	RecordID        string                `json:"record_id"`
	Variant         string                `json:"variant"`
	ModelVersionID  uuid.UUID             `json:"model_version_id"`
	PredictedLabel  string                `json:"predicted_label"`
	Score           float64               `json:"score"`
	Confidence      float64               `json:"confidence"`
	Contributions   []FeatureContribution `json:"contributions"`
}

type RealtimePredictionRequest struct {
	Inputs  []json.RawMessage `json:"inputs,omitempty"`
	Explain bool              `json:"explain,omitempty"`
}

type RealtimePredictionResponse struct {
	DeploymentID uuid.UUID          `json:"deployment_id"`
	Outputs      []PredictionOutput `json:"outputs"`
	PredictedAt  time.Time          `json:"predicted_at"`
}

type CreateBatchPredictionRequest struct {
	DeploymentID      uuid.UUID         `json:"deployment_id"`
	Records           []json.RawMessage `json:"records,omitempty"`
	OutputDestination *string           `json:"output_destination"`
}

type BatchPredictionJob struct {
	ID                uuid.UUID          `json:"id"`
	DeploymentID      uuid.UUID          `json:"deployment_id"`
	Status            string             `json:"status"`
	RecordCount       int64              `json:"record_count"`
	OutputDestination *string            `json:"output_destination"`
	Outputs           []PredictionOutput `json:"outputs"`
	CreatedAt         time.Time          `json:"created_at"`
	CompletedAt       *time.Time         `json:"completed_at"`
}

type ListBatchPredictionsResponse struct {
	Data []BatchPredictionJob `json:"data"`
}
